File: app/src/main/java/com/billpayments/core/theme/AppTheme.kt
package com.billpayments.core.theme

import android.app.Activity
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.SideEffect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalView
import androidx.core.view.WindowCompat
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.billpayments.R
import com.billpayments.features.settings.data.AppThemeMode

val BrandBlue = Color(0xFF11448A)
val BrandBlueLight = Color(0xFF3161AD)
val BrandYellow = Color(0xFFFACC15)
val BrandYellowSoft = Color(0xFFFFCA49)

private val Rubik = FontFamily(Font(R.font.rubik))

private val LightColors = lightColorScheme(
    primary = BrandBlue,
    onPrimary = Color.White,
    secondary = BrandYellow,
    onSecondary = BrandBlue,
    tertiary = BrandYellowSoft,
    onTertiary = BrandBlue,
)

private val DarkColors = darkColorScheme(
    primary = BrandBlueLight,
    onPrimary = Color.White,
    secondary = BrandYellow,
    onSecondary = BrandBlue,
    tertiary = BrandYellowSoft,
    onTertiary = BrandBlue,
)

private fun rubikTypography(): Typography {
    val base = Typography()
    return Typography(
        displayLarge = base.displayLarge.copy(fontFamily = Rubik),
        displayMedium = base.displayMedium.copy(fontFamily = Rubik),
        displaySmall = base.displaySmall.copy(fontFamily = Rubik),
        headlineLarge = base.headlineLarge.copy(fontFamily = Rubik, fontWeight = FontWeight.W700),
        headlineMedium = base.headlineMedium.copy(fontFamily = Rubik, fontWeight = FontWeight.W700),
        headlineSmall = base.headlineSmall.copy(fontFamily = Rubik, fontWeight = FontWeight.W700),
        titleLarge = base.titleLarge.copy(fontFamily = Rubik, fontWeight = FontWeight.W700),
        titleMedium = base.titleMedium.copy(fontFamily = Rubik),
        titleSmall = base.titleSmall.copy(fontFamily = Rubik),
        bodyLarge = base.bodyLarge.copy(fontFamily = Rubik),
        bodyMedium = base.bodyMedium.copy(fontFamily = Rubik),
        bodySmall = base.bodySmall.copy(fontFamily = Rubik),
        labelLarge = base.labelLarge.copy(fontFamily = Rubik),
        labelMedium = base.labelMedium.copy(fontFamily = Rubik),
        labelSmall = base.labelSmall.copy(fontFamily = Rubik),
    )
}

private val AppShapes = Shapes(
    small = RoundedCornerShape(8.dp),
    medium = RoundedCornerShape(12.dp),
    large = RoundedCornerShape(12.dp),
)

@Composable
fun BillPaymentsTheme(themeMode: AppThemeMode, content: @Composable () -> Unit) {
    val dark = when (themeMode) {
        AppThemeMode.System -> isSystemInDarkTheme()
        AppThemeMode.Light -> false
        AppThemeMode.Dark -> true
    }
    val view = LocalView.current
    if (!view.isInEditMode) {
        SideEffect {
            val window = (view.context as Activity).window
            val controller = WindowCompat.getInsetsController(window, view)
            controller.isAppearanceLightStatusBars = !dark
            controller.isAppearanceLightNavigationBars = !dark
        }
    }
    MaterialTheme(
        colorScheme = if (dark) DarkColors else LightColors,
        typography = rubikTypography(),
        shapes = AppShapes,
        content = content,
    )
}

File: lib/core/theme/app_theme.dart
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';

abstract final class AppTheme {
  static const brandBlue = Color(0xFF11448A);
  static const brandBlueLight = Color(0xFF3161AD);
  static const brandYellow = Color(0xFFFACC15);
  static const brandYellowSoft = Color(0xFFFFCA49);

  static ThemeData get light => _build(_lightScheme);

  static ThemeData get dark => _build(_darkScheme);

  static ColorScheme get _lightScheme =>
      ColorScheme.fromSeed(seedColor: brandBlue).copyWith(
        primary: brandBlue,
        onPrimary: Colors.white,
        secondary: brandYellow,
        onSecondary: brandBlue,
        tertiary: brandYellowSoft,
        onTertiary: brandBlue,
      );

  static ColorScheme get _darkScheme => ColorScheme.fromSeed(
        seedColor: brandBlue,
        brightness: Brightness.dark,
      ).copyWith(
        primary: brandBlueLight,
        onPrimary: Colors.white,
        secondary: brandYellow,
        onSecondary: brandBlue,
        tertiary: brandYellowSoft,
        onTertiary: brandBlue,
      );

  static ThemeData _build(ColorScheme scheme) {
    final base = ThemeData(colorScheme: scheme);
    return base.copyWith(
      textTheme: _rubikTextTheme(base.textTheme),
      appBarTheme: _appBarTheme(scheme),
      filledButtonTheme: FilledButtonThemeData(style: _buttonStyle),
      elevatedButtonTheme: ElevatedButtonThemeData(style: _buttonStyle),
      cardTheme: base.cardTheme.copyWith(shape: _roundedBorder(12)),
      chipTheme: base.chipTheme.copyWith(shape: _roundedBorder(8)),
    );
  }

  static TextTheme _rubikTextTheme(TextTheme base) {
    final rubik = GoogleFonts.rubikTextTheme(base);
    return rubik.copyWith(
      headlineLarge: rubik.headlineLarge?.copyWith(fontWeight: FontWeight.w700),
      headlineMedium:
          rubik.headlineMedium?.copyWith(fontWeight: FontWeight.w700),
      headlineSmall: rubik.headlineSmall?.copyWith(fontWeight: FontWeight.w700),
      titleLarge: rubik.titleLarge?.copyWith(fontWeight: FontWeight.w700),
    );
  }

  static AppBarTheme _appBarTheme(ColorScheme scheme) {
    if (scheme.brightness == Brightness.dark) return const AppBarTheme();
    return AppBarTheme(
      backgroundColor: scheme.primary,
      foregroundColor: scheme.onPrimary,
    );
  }

  static ButtonStyle get _buttonStyle =>
      ButtonStyle(shape: WidgetStatePropertyAll(_roundedBorder(8)));

  static RoundedRectangleBorder _roundedBorder(double radius) =>
      RoundedRectangleBorder(borderRadius: BorderRadius.circular(radius));
}
